### PR TITLE
Address Memory hotplug limitation

### PR DESCRIPTION
**Pull request by egernst**
If we're using ACPI hotplug for memory, there's a limitation on the
amount of memory which can be hotplugged at a single time.

During hotplug, we'll allocate memory for the memmap for each page,
resulting in a 64 byte per 4KiB page allocation. As an example, hotplugging 12GiB
of memory requires ~192 MiB of *free* memory, which is about the limit
we should expect for an idle 256 MiB guest (conservative heuristic of 75%
of provided memory).

From experimentation, at pod creation time we can reliably add 48 times
what is provided to the guest. (a factor of 48 results in using 75% of
provided memory for hotplug). Using prior example of a guest with 256Mi
RAM, 256 Mi * 48 = 12 Gi; 12GiB is upper end of what we should expect
can be hotplugged successfully into the guest.

Note: It isn't expected that we'll need to hotplug large amounts of RAM
after workloads have already started -- container additions are expected
to occur first in pod lifecycle. Based on this, we expect that provided
memory should be freely available for hotplug.

If virtio-mem is being utilized, there isn't such a limitation - we can
hotplug the max allowed memory at a single time.

Fixes: https://github.com/kata-containers/kata-containers/issues/4847

Signed-off-by: Eric Ernst <eric_ernst@apple.com>

### CODE REVIEW

File: src/runtime/virtcontainers/sandbox.go [hunk 1]
```diff
@@ -2012,9 +2012,60 @@ func (s *Sandbox) updateResources(ctx context.Context) error {
...
 
-	// Update Memory
-	s.Logger().WithField("memory-sandbox-size-byte", sandboxMemoryByte).Debugf("Request to hypervisor to update memory")
+	// Update Memory --
+	// If we're using ACPI hotplug for memory, there's a limitation on the amount of memory which can be hotplugged at a single time.
+	// We must have enough free memory in the guest kernel to cover 64bytes per (4KiB) page of memory added for mem_map.
+	// See https://github.com/kata-containers/kata-containers/issues/4847 for more details.
+	// For typical pod lifecycle, we expect that each container is added then we start the workloads. based on this, we'll "assume" that majority
+	// of the guest memory is readily available. From experimentation, we see that we can add approximately 48 times what is already provided to
+	// the guest workload. For example, a 256 MiB guest should be able to accomodate hotplugging 12 GiB  of memory.
```
Comment (mcastelino) [inline src/runtime/virtcontainers/sandbox.go:2021]: Can this be impacted by amount of memory currently in use within the guest. In our case we do the hot plug really early, so it should not be an issue. But if we hot plug later we may need to take that into consideration.

Comment (egernst) [reply src/runtime/virtcontainers/sandbox.go:2021]: Yes, for sure. If you later try to add a very large container, and the others are using near their limit, there's a good chance the hot plug will fail.  Given expected pod lifecycle, I don't expect this scenario.  This PR should address the existing common scenario, but we're using a heuristic here.  Ideally, when triggering the hot plug event we'd have a better signal to see if we were able to actually add the memory or not.  This is disjoint today.

**Comment by egernst** [comment]
I noticed that the updateResources unit tests are very much lacking, and I think broken. keeping this PR WIP as I verify the tests  behavior. Works will locally e2e :)

**Review by cmaf — COMMENTED** [review]
@egernst if I am understanding it correctly, using an`acpiMemoryHotplugFactor` with a higher number (the ones from the table in the issue vs the example in the comments) would result in a higher failure rate, vs the conservatively computed one which is less likely to fail at the cost of it potentially requiring an additional hotplug?

Is there a reason we shouldn't just compute the maximum hotpluggable memory based on the known memory overhead and the current size (or expected free memory) rather than a heuristic based on a 256 MB guest? (Asking based on the factors you calculated from different guest memory, seen in the issue)

**Comment by egernst** [comment]
Hey Chelsea -- thanks for the review help. Indeed, 48 lined up closely to the 256 MiB case I described in the issue, and aligned with the expected overhead based on insights from David Hildenbrand.

Realistically, if an end user is adding a (very) large container, they'll need to break it up into 2 individual hot plugs either way. I'd rather rely on 75% of the existing RAM rather than pushing for a more aggressive number to: (1) keep things simple, (2) be more conservative.

This'll be more foolproof when we can rely on CRI passing the total sizing information so we can avoid the initial hot plug in the first place, but for now, I want to have a workaround in place for folks who aren't on >=1.23 version of Kubernetes.

**Comment by cmaf** [comment]
@egernst makes sense, lgtm

**Comment by bergwolf** [comment]
@egernst IIUC, it also depends on if the guest has hugepage enabled or not. So it is not always the same linear calculation.

File: src/runtime/virtcontainers/sandbox.go [hunk 2]
```diff
@@ -77,6 +77,14 @@ const (
 
 	// Restricted permission for shared directory managed by virtiofs
 	sharedDirMode = os.FileMode(0700) | os.ModeDir
+
+	// hotplug factor indicates how much memory can be hotplugged relative to the amount of
+	// RAM provided to the guest. This is a conservative heuristic based on neededing 64 bytes per
```
Comment (littlejawa) [inline src/runtime/virtcontainers/sandbox.go:82]: nit: s/neededing/needing/

File: src/runtime/virtcontainers/sandbox.go [hunk 3]
```diff
@@ -2012,9 +2020,60 @@ func (s *Sandbox) updateResources(ctx context.Context) error {
 	}
 	s.Logger().Debugf("Sandbox CPUs: %d", newCPUs)
 
-	// Update Memory
-	s.Logger().WithField("memory-sandbox-size-byte", sandboxMemoryByte).Debugf("Request to hypervisor to update memory")
+	// Update Memory --
+	// If we're using ACPI hotplug for memory, there's a limitation on the amount of memory which can be hotplugged at a single time.
+	// We must have enough free memory in the guest kernel to cover 64bytes per (4KiB) page of memory added for mem_map.
+	// See https://github.com/kata-containers/kata-containers/issues/4847 for more details.
+	// For typical pod lifecycle, we expect that each container is added then we start the workloads. based on this, we'll "assume" that majority
```
Comment (littlejawa) [inline src/runtime/virtcontainers/sandbox.go:2027]: nit:
- s/then/when/
- s/based/Based/

File: src/runtime/virtcontainers/sandbox.go [hunk 4]
```diff
@@ -2012,9 +2020,60 @@ func (s *Sandbox) updateResources(ctx context.Context) error {
...
+	// of the guest memory is readily available. From experimentation, we see that we can add approximately 48 times what is already provided to
+	// the guest workload. For example, a 256 MiB guest should be able to accommodate hotplugging 12 GiB  of memory.
```
Comment (littlejawa) [inline src/runtime/virtcontainers/sandbox.go:2029]: nit: additional space between "GiB" and "of"

File: src/runtime/virtcontainers/sandbox.go [hunk 5]
```diff
@@ -2012,9 +2020,60 @@ func (s *Sandbox) updateResources(ctx context.Context) error {
...
+
+	hconfig := s.hypervisor.HypervisorConfig()
+
+	for {
+		currentMemoryMB := s.hypervisor.GetTotalMemoryMB(ctx)
+
+		maxhotPluggableMemoryMB := currentMemoryMB * acpiMemoryHotplugFactor
+
+		// In the case of virtio-mem, we don't have a restriction on how much can be hotplugged at
+		// a single time. As a result, the max hotpluggable is only limited my the maximum memory size
```
Comment (littlejawa) [inline src/runtime/virtcontainers/sandbox.go:2044]: nit: s/my/by/

**Review by littlejawa — APPROVED** [review]
Found some typos in comments, but no issue in the code.
LGTM

**Comment by egernst** [comment]
I'm seeing some issues where the guest becomes unresponsive when carrying out the on-lining which may be related to the failure that's seen in azure for CI.  Debugging now....

**Comment by egernst** [comment]
Seeing:
```
not ok 2 Running within memory constraints
# (in test file k8s-memory.bats, line 49)
#   `kubectl wait --for=condition=Ready --timeout=$timeout pod "$pod_name"' failed
```

**Comment by egernst** [comment]
🤦🏻  - I identified original patchset had issue when the added memory wasn't 128MiB aligned, the delta between what we have and what we'd want is negative, but we comparing unsigned.  Fixed it up and re-pushed.

/test

**Comment by egernst** [comment]
> @egernst IIUC, it also depends on if the guest has hugepage enabled or not. So it is not always the same linear calculation.

Indeed, if you're using 2MB pages, assuming its still 64B per page, you'd have a much different number. Page size can vary with huge pages -- I wonder if for a first pass its easiest to just use the most conservative number (4KiB) for pages? I can foresee this getting complicated if we want to find the "best" amount allowed for hot plug for each potential page size?

**Review by fengwang666 — APPROVED** [review]
LGTM! Thanks!

**Comment by bergwolf** [comment]
> Indeed, if you're using 2MB pages, assuming its still 64B per page, you'd have a much different number. Page size can vary with huge pages -- I wonder if for a first pass its easiest to just use the most conservative number (4KiB) for pages? I can foresee this getting complicated if we want to find the "best" amount allowed for hot plug for each potential page size?

Yeah, I can live with it as a first pass. In the longer term, we should return the hugepage information in `GuestDetailsResponse`

**Review by bergwolf — APPROVED** [review]
lgtm, thanks!

**Pull request merged**